src/nifs/protogalaxy/poly/folded_trace.rs

```diff
@@ -0,0 +1,180 @@
+use std::iter;
+
+use ff::PrimeField;
+use rayon::prelude::*;
+
+use crate::{
+    plonk::{GetChallenges, GetWitness, PlonkWitness},
+    polynomial::lagrange,
+    util::MultiCartesianProduct,
+};
+
+pub(crate) struct FoldedTrace<F: PrimeField> {
+    witness: PlonkWitness<F>,
+    challenges: Vec<F>,
+}
+
+impl<F: PrimeField> FoldedTrace<F> {
+    pub(crate) fn new(
+        points_for_fft: &[F],
+        accumulator: impl Sync + GetChallenges<F> + GetWitness<F>,
+        traces: &[(impl Sync + GetChallenges<F> + GetWitness<F>)],
+    ) -> Box<[Self]> {
+        let folded_witnesses_collection = fold_witnesses(points_for_fft, &accumulator, traces);
+        let folded_challenges_collection =
+            fold_plonk_challenges(points_for_fft, &accumulator, traces);
+
+        folded_witnesses_collection
+            .into_iter()
+            .zip(folded_challenges_collection)
+            .map(|(witness, challenges)| Self {
+                witness,
+                challenges,
+            })
+            .collect()
+    }
+}
+impl<F: PrimeField> GetChallenges<F> for FoldedTrace<F> {
+    fn get_challenges(&self) -> &[F] {
+        &self.challenges
+    }
+}
+impl<F: PrimeField> GetWitness<F> for FoldedTrace<F> {
+    fn get_witness(&self) -> &[Vec<F>] {
+        &self.witness.W
+    }
+}
+
+/// For each `X` we must perform the operation of sum all all matrices [`PlonkWitness`] with
+/// coefficients taken from [`lagrange::iter_eval_lagrange_polynomials_for_cyclic_group`]
+///
+/// Since the number of rows is large, we do this in one pass, counting the points for each
+/// challenge at each iteration, and laying them out in separate [`PlonkWitness`] at the end.
+fn fold_witnesses<F: PrimeField>(
+    X_challenges: &[F],
+    accumulator: &(impl GetWitness<F> + Sync),
+    witnesses: &[impl Sync + GetWitness<F>],
+) -> Vec<PlonkWitness<F>> {
+    let log_n = (witnesses.len() + 1).next_power_of_two().ilog2();
+
+    let lagrange_poly_by_challenge = X_challenges
+        .iter()
+        .map(|X| {
+            lagrange::iter_eval_lagrange_polynomials_for_cyclic_group(*X, log_n)
+                .collect::<Box<[_]>>()
+        })
+        .collect::<Box<[_]>>();
+
+    let empty_witness = accumulator
+        .get_witness()
+        .iter()
+        .map(|row| vec![F::ZERO; row.len()]) // TODO Use memory allocation with zeroes, for not iterate over `row.len()`
+        .collect::<Vec<_>>();
+
+    // TODO Create on the fly to avoid multiple rows iterations
+    let mut result_matrix_by_challenge = vec![
+        PlonkWitness {
+            W: empty_witness.clone(),
```

Comment:
Yes, you are right, I can remove that particular clone, but you have to realize that there will still be `X_challenges.len` clones of that vector here, because it's inside `vec!` macro